<a href="https://colab.research.google.com/github/tvjoseph/Packt/blob/master/Copy_of_Chapter_13_Unbalanced_Data_sets.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# JMJPFU
# 28-July-2019

This notbook has all the exercises for Chapter 13

reference

https://www.kaggle.com/janiobachmann/credit-fraud-dealing-with-imbalanced-datasets

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
# Loading the necessary library files
import pandas as pd

In [5]:
# Loading data from the drive

# Please change the filename as per the location where the file is stored
filename = '/content/drive/My Drive/Packt_Colab/bank-full.csv'
# Loading the data u'sing pandas

bankData = pd.read_csv(filename,sep=";")
bankData.head()

age           job  marital  education  ... pdays  previous poutcome   y
0   58    management  married   tertiary  ...    -1         0  unknown  no
1   44    technician   single  secondary  ...    -1         0  unknown  no
2   33  entrepreneur  married  secondary  ...    -1         0  unknown  no
3   47   blue-collar  married    unknown  ...    -1         0  unknown  no
4   33       unknown   single    unknown  ...    -1         0  unknown  no

[5 rows x 17 columns]

In [6]:
bankData.shape

(45211, 17)

In [7]:
(bankData['y'].value_counts()/len(bankData) ) * 100

no     88.30152
yes    11.69848
Name: y, dtype: float64

**Feature engineering steps**

Let us now do some feature engineering to the data. First we will scale the numerical data and then convert the ordinal data to 
dummy data

In [0]:
from sklearn.preprocessing import StandardScaler, RobustScaler

std_scaler = StandardScaler()
rob_scaler = RobustScaler()

# Converting each of the columns to scaled version

bankData['ageScaled'] = rob_scaler.fit_transform(bankData['age'].values.reshape(-1,1))
bankData['balScaled'] = rob_scaler.fit_transform(bankData['balance'].values.reshape(-1,1))
bankData['durScaled'] = rob_scaler.fit_transform(bankData['duration'].values.reshape(-1,1))

# Dropping the original columns

bankData.drop(['age','balance','duration'], axis=1, inplace=True)

In [9]:
bankData.head()

job  marital  education  ... ageScaled balScaled durScaled
0    management  married   tertiary  ...  1.266667  1.250000  0.375000
1    technician   single  secondary  ...  0.333333 -0.308997 -0.134259
2  entrepreneur  married  secondary  ... -0.400000 -0.328909 -0.481481
3   blue-collar  married    unknown  ...  0.533333  0.780236 -0.407407
4       unknown   single    unknown  ... -0.400000 -0.329646  0.083333

[5 rows x 17 columns]

In [0]:
# Converting all the categorical variables to dummy variables
bankCat = pd.get_dummies(bankData[['job','marital','education','default','housing','loan','contact','month','poutcome']])

In [11]:
# Seperating the numerical data
bankNum = bankData[['ageScaled','balScaled','day','durScaled','campaign','pdays','previous']]
bankNum.shape

(45211, 7)

In [12]:
# Merging with the original data frame
# Preparing the X variables
X = pd.concat([bankCat, bankNum], axis=1)
print(X.shape)
# Preparing the Y variable
Y = bankData['y']
print(Y.shape)
X.head()

(45211, 51)
(45211,)


job_admin.  job_blue-collar  job_entrepreneur  ...  campaign  pdays  previous
0           0                0                 0  ...         1     -1         0
1           0                0                 0  ...         1     -1         0
2           0                0                 1  ...         1     -1         0
3           0                1                 0  ...         1     -1         0
4           0                0                 0  ...         1     -1         0

[5 rows x 51 columns]

In [13]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
# Splitting the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3, random_state=0)
# Defining the LogisticRegression function
bankModel = LogisticRegression()
bankModel.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [14]:
pred = bankModel.predict(X_test)
print('Accuracy of Logisticr regression model prediction on test set: {:.2f}'.format(bankModel.score(X_test, y_test)))

# Confusion Matrix for the model
from sklearn.metrics import confusion_matrix
confusionMatrix = confusion_matrix(y_test, pred)
print(confusionMatrix)

Accuracy of Logisticr regression model prediction on test set: 0.90
[[11639   330]
 [ 1038   557]]


**Undersampling Method.**

In the random undersampling method, we down sample the majority class to the same amount as the minority class to make the data set balanced. Let us see how we can achieve that

In this method we first identify the count of  the  minority cases and then undersample the majority cases to be the same as minority cases. 




In [17]:
# let us first join the train_x and train_y for ease of operation

trainData = pd.concat([X_train,y_train],axis=1)
trainData.head()

job_admin.  job_blue-collar  job_entrepreneur  ...  pdays  previous   y
24951           0                0                 0  ...     -1         0  no
22129           0                1                 0  ...     -1         0  no
986             1                0                 0  ...     -1         0  no
14999           0                0                 0  ...     -1         0  no
4027            0                0                 0  ...     -1         0  no

[5 rows x 52 columns]

In [19]:
# Finding the indexes of the sample data set where the 
ind = trainData[trainData['y']=='yes'].index
len(ind)

3694

In [20]:
ind

Int64Index([33832, 31327, 39079, 41159, 33852, 44106, 43319, 26193, 34565,
            16367,
            ...
            31055,  6021, 43890, 31230, 42968, 24152, 39512, 15430, 41993,
            42613],
           dtype='int64', length=3694)

In [24]:
# Let us first seperate the 'Yes' cases
minData = trainData.loc[ind]
minData.head()



job_admin.  job_blue-collar  job_entrepreneur  ...  pdays  previous    y
33832           1                0                 0  ...     -1         0  yes
31327           0                0                 0  ...     -1         0  yes
39079           0                0                 0  ...    347         2  yes
41159           1                0                 0  ...     84         3  yes
33852           0                1                 0  ...    349         1  yes

[5 rows x 52 columns]

In [22]:
minData.shape

(3694, 52)

In [36]:
# Finding indexes of majority class
ind1 = trainData[trainData['y']=='no'].index
print(len(ind1))
# Taking the seperate data set for majority class
majData = trainData.loc[ind1]
print(majData.shape)
majData.head()


27953
(27953, 52)


job_admin.  job_blue-collar  job_entrepreneur  ...  pdays  previous   y
24951           0                0                 0  ...     -1         0  no
22129           0                1                 0  ...     -1         0  no
986             1                0                 0  ...     -1         0  no
14999           0                0                 0  ...     -1         0  no
4027            0                0                 0  ...     -1         0  no

[5 rows x 52 columns]

In [37]:
# Take a random sample equal to length of the minority class to make the data set balanced

majSample = majData.sample(n=len(ind))
print(majSample.shape)
majSample.head()



(3694, 52)


job_admin.  job_blue-collar  job_entrepreneur  ...  pdays  previous   y
8179            0                1                 0  ...     -1         0  no
4305            0                0                 0  ...     -1         0  no
30219           1                0                 0  ...     -1         0  no
4002            0                1                 0  ...     -1         0  no
34813           0                1                 0  ...    338         2  no

[5 rows x 52 columns]

In [41]:
# Concatinating both data sets and then shuffling the data set

balData = pd.concat([minData,majSample],axis = 0)
print('balanced data set shape',balData.shape)

# Shuffling the data set

from sklearn.utils import shuffle

balData = shuffle(balData)
balData.head()

balanced data set shape (7388, 52)


job_admin.  job_blue-collar  job_entrepreneur  ...  pdays  previous    y
42567           0                0                 0  ...     -1         0  yes
16193           0                0                 0  ...     -1         0   no
18505           0                0                 0  ...     -1         0   no
12303           0                0                 0  ...     -1         0   no
15860           0                0                 0  ...     -1         0  yes

[5 rows x 52 columns]

In [52]:
(balData['y'].value_counts()/len(balData) ) * 100

yes    50.0
no     50.0
Name: y, dtype: float64

In [50]:
# Making the new X_train and y_train

X_trainNew = balData.iloc[:,0:51]
X_trainNew.head()

y_trainNew = balData['y']
y_trainNew.head()

42567    yes
16193     no
18505     no
12303     no
15860    yes
Name: y, dtype: object

In [54]:

# Defining the LogisticRegression function
bankModel1 = LogisticRegression()
bankModel1.fit(X_trainNew, y_trainNew)

# Predicting on the test
pred = bankModel1.predict(X_test)
print('Accuracy of Logisticr regression model prediction on test set for balanced data set: {:.2f}'.format(bankModel1.score(X_test, y_test)))



/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Accuracy of Logisticr regression model prediction on test set for balanced data set: 0.84
[[10083  1886]
 [  290  1305]]


In [55]:
# Confusion Matrix for the model
from sklearn.metrics import confusion_matrix
confusionMatrix = confusion_matrix(y_test, pred)
print(confusionMatrix)

[[10083  1886]
 [  290  1305]]


In [56]:
from sklearn.metrics import classification_report
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

          no       0.97      0.84      0.90     11969
         yes       0.41      0.82      0.55      1595

    accuracy                           0.84     13564
   macro avg       0.69      0.83      0.72     13564
weighted avg       0.91      0.84      0.86     13564



Let us now try to analyse our results which we have got now. Let us compare it with the logistic regression model which we built in chapter 3. if we recall in that chapter we had the problem of model being biased towards the majority class with a very low recall for the 'yes' cases. 

Now by balancing the data set we have seen that the recall for the minority class has improved tremendously from a low of 0.35 to around 0.82. 

However we can see that our overall accuracy has taken a hit. From a high of around 90% it has come down to around 84%. This is within expected lines because the absolute data set was greatly reduced from around 31,000 examples to around 7500 examples. So the accuracy has taken a hit. One major area where there accuracy has taken a hit is the amount of false positives i.e those 'No' cases which were wrongly predicted as 'yes'. Looking this at from a business perspective, this is a much better result than the one we got in chapter 3. We were able to identify more customers who were likely to buy a term deposits, which can potentially result in better conversion rate. However the flip side of this is that the sales team will also have to sent lot of time on customers who are unlikely to buy term deposits.

Ideally we would want the quadrant 2 and 3 come down in favour of other two quadrants. Let us now attempt other methods and see how it fares